hugo/content/3-queries/introduction.md

```diff
@@ -0,0 +1,169 @@
++++
+next = "/3-queries/operators"
+prev = "/3-queries"
+title = "Introduction"
+toc = true
+weight = 1
+
++++
+
+Mangrove provides a **query builder** with static type-checking for constructing MongoDB queries.
+This allows you to compare documents using familiar C++ syntax using operators such as `>` and `!=`.
+These query objects are automatically converted to BSON, and can be passed either to the Mangrove Model's `find` and `find_one` methods, to the Mangrove collection wrapper, or to the query methods in the Mongo C++ Driver.
+
+The query builder helps you avoid JSON syntax errors, such as forgetting a closing curly brace, and will catch other errors at compile time --- so you don't have to wait until your queries return errors in production (or silently return incorrect results!) to catch mistakes.
+
+Mangrove currently supports all the {{% a_blank "query operators" "https://docs.mongodb.com/v3.2/reference/operator/query/" %}} as of MongoDB 3.2,
+except for `$type`, `$where`, and geospatial queries. If you need to use these operators, create the BSON for the query as you {{% a_blank "normally would" "https://github.com/mongodb/mongo-cxx-driver/wiki/Handling-BSON-in-the-new-driver" %}} using the MongoDB C++ driver.
+
+Consider the following class that represents a user of some website:
+```cpp
+class User {
+public:
+    std::string username;
+    std::int32_t age;
+
+    MANGROVE_MAKE_KEYS(User, MANGROVE_NVP(username), MANGROVE_NVP(age));
+};
+```
+
+To retrieve the users in the database that are over 21 years of age, one would use the following query:
+
+```cpp
+// { age: {$gte: 21} }
+auto results = collection.find(MANGROVE_KEY(User::age) >= 21);
+```
+
+Where `collection` is a `mongocxx::collection` object from the C++ Driver.
+The different parts of this code will be explained in more detail below,
+but note the `MANGROVE_KEY` macro that is used to refer to individual fields in the document.
+Then, Mangrove uses {{% a_blank "operator overloading" "http://en.cppreference.com/w/cpp/language/operators" %}} to allow the field to be compared to `21` with the `>=` operator.
+This constructs a query object, which is implicitly cast to a BSON value when passed into `mongocxx::collection::find()`.
+
+{{% notice warning %}}
+The query objects should only be used as *temporary objects* --- that is, constructed and passed directly into another function, such as `find()`.
+Attempting to store these queries in a variable and use them later may produce undefined results.
+This is because the query expressions store their arguments by reference to avoid putting too much data on the stack.
+{{% /notice %}}
+
+## Specifying fields
+
+You've already seen one way of referring to fields, the `MANGROVE_KEY` macro.
+This macro takes a class member specified as `Base::member`, and creates a name-value pair that can be used in queries.
```

Comment:
That sounds good to me. At least in the documentation I can change mentions of "name-value pairs" to something like "fields" or "field objects", and later we can change the various `nvp` classes in the code.